### PR TITLE
GHST: Add support for GPS Telemetry

### PR DESCRIPTION
**Pull request by tonycake**
Add support for the basic GPS telemetry values when using the GHST protocol.

Initial GHST driver handled battery status telemetry, but not GPS telemetry. This PR contains a relatively simple (safe) change which adds pass-through of PX4's GPS position to the Ghost receiver. 

Tested on a H743-SLIM (which requires #define RC_SERIAL_SINGLEWIRE in board_config.h to enable telemetry). 
Reviewers... should this #define be added to the H743-SLIM board definition? it does mean (I believe) that anyone running this board with the Ghost receiver hooked to a UART Rx pin would need to rewire it to a Tx pin (as is the case for Betaflight, iNav, etc. )

### CODE REVIEW

**Comment by Igor-Misic** [comment]
Hey @tonycake congrats on the first PX4 contribution :tada: It is nice to have you here.

**Comment by tonycake** [comment]
> Hey @tonycake congrats on the first PX4 contribution 🎉 It is nice to have you here.

Just getting started Igor :-) :-)

Any thoughts on the RC_SERIAL_SINGLEWIRE comment? Kindof a shame to have to recompile to enable telemetry on boards that don't have it defined.... unless there is a trick to enable it at run time that I didn't fine.

**Comment by Igor-Misic** [comment]
> > Hey @tonycake congrats on the first PX4 contribution tada It is nice to have you here.
> 
> Just getting started Igor :-) :-)
> 
> Any thoughts on the RC_SERIAL_SINGLEWIRE comment? Kindof a shame to have to recompile to enable telemetry on boards that don't have it defined.... unless there is a trick to enable it at run time that I didn't fine.

If you trace down `RC_SERIAL_SINGLEWIRE` you will see that it ends into this sbus config function that enables single wire.
https://github.com/PX4/PX4-Autopilot/blob/main/src/lib/rc/sbus.cpp#L155

This function should be refactored into general purpose one since now there are more Receivers that use singlewire. After that, it could be used for GHST to automatically check Rx and Tx pin to see where is Ghost connected (or fixed to Tx, whatever you like more).

File: src/drivers/rc_input/ghst_telemetry.cpp [hunk 1]
```diff
@@ -93,3 +101,39 @@ bool GHSTTelemetry::send_battery_status()
...
+
+	if (!_vehicle_gps_position_sub.update(&vehicle_gps_position)) {
+		return false;
+	}
+
+	uint16_t ground_speed = (uint16_t)(vehicle_gps_position.vel_d_m_s / 3.6f * 10.f);
+	uint16_t ground_course = (uint16_t)(math::degrees(vehicle_gps_position.cog_rad) * 100.f);
+	uint8_t num_sats = vehicle_gps_position.satellites_used;
+
+	// TBD: Can these be computed in a RC telemetry driver?
```
Comment (Igor-Misic) [inline src/drivers/rc_input/ghst_telemetry.cpp:132]: Can this help? 
https://github.com/PX4/PX4-Autopilot/blob/main/msg/home_position.msg

File: src/drivers/rc_input/ghst_telemetry.hpp [hunk 2]
```diff
@@ -44,6 +44,9 @@
 
 #include <uORB/Subscription.hpp>
 #include <uORB/topics/battery_status.h>
+#include <uORB/topics/vehicle_attitude.h>
```
Comment (Igor-Misic) [inline src/drivers/rc_input/ghst_telemetry.hpp:47]:
```suggestion
```
Not used.

File: src/drivers/rc_input/ghst_telemetry.hpp [hunk 3]
```diff
@@ -44,6 +44,9 @@
...
+#include <uORB/topics/sensor_gps.h>
+#include <uORB/topics/vehicle_status.h>
```
Comment (Igor-Misic) [inline src/drivers/rc_input/ghst_telemetry.hpp:49]:
```suggestion
```

Not used.

File: src/lib/rc/ghst.cpp [hunk 4]
```diff
@@ -385,3 +398,34 @@ bool ghst_send_telemetry_battery_status(int uart_fd, uint16_t voltage_in_10mV,
...
+	write_frame_header(buf, offset, ghstTelemetryType::gpsPrimary, GHST_FRAME_PAYLOAD_SIZE_TELEMETRY);
+	write_uint32_t(buf, offset, latitude);
+	write_uint32_t(buf, offset, longitude);
+	write_uint16_t(buf, offset, altitude);
+	write_frame_crc(buf, offset, sizeof(buf));
+
+	return write(uart_fd, buf, offset) == offset;
+}
+
+bool ghst_send_telemetry_gps2_status(int uart_fd, uint16_t groundSpeed, uint16_t groundCourse, uint8_t numSats,
```
Comment (Igor-Misic) [inline src/lib/rc/ghst.cpp:415]:
```suggestion
bool ghst_send_telemetry_gps2_status(int uart_fd, uint16_t ground_speed, uint16_t ground_course, uint8_t num_sats,
```

File: src/lib/rc/ghst.cpp [hunk 5]
```diff
@@ -385,3 +398,34 @@ bool ghst_send_telemetry_battery_status(int uart_fd, uint16_t voltage_in_10mV,
...
+				     uint16_t homeDist, uint16_t homeDir, uint8_t flags)
```
Comment (Igor-Misic) [inline src/lib/rc/ghst.cpp:416]:
```suggestion
				     uint16_t home_dist, uint16_t home_dir, uint8_t flags)
```

File: src/lib/rc/ghst.hpp [hunk 6]
```diff
@@ -139,4 +141,27 @@ __EXPORT bool ghst_parse(const uint64_t now, const uint8_t *frame, unsigned len,
...
+ * @param uart_fd UART file descriptor
+ * @param groundSpeed Ground Speed [1 km/h]
+ * @param groundCourse Ground Course [1e-7 degrees]
+ * @param numSats GPS Satellite count
+ * @param homeDist Distance to Home [10 m]
+ * @param homeDir Direction to Home [1e-7 degrees]
+ * @param flags GPS Flags
+ * @return true on success
+ */
+__EXPORT bool ghst_send_telemetry_gps2_status(int uart_fd, uint16_t groundSpeed, uint16_t groundCourse, uint8_t numSats,
```
Comment (Igor-Misic) [inline src/lib/rc/ghst.hpp:165]:
```suggestion
__EXPORT bool ghst_send_telemetry_gps2_status(int uart_fd, uint16_t ground_speed, uint16_t ground_course, uint8_t num_sats,
```

File: src/lib/rc/ghst.hpp [hunk 7]
```diff
@@ -139,4 +141,27 @@ __EXPORT bool ghst_parse(const uint64_t now, const uint8_t *frame, unsigned len,
...
+		uint16_t homeDist, uint16_t homeDir, uint8_t flags);
```
Comment (Igor-Misic) [inline src/lib/rc/ghst.hpp:166]:
```suggestion
		uint16_t home_dist, uint16_t home_dir, uint8_t flags);
```

**Review by Igor-Misic — COMMENTED** [review]
This can go in after another round of code cleaning. Changes added a minor increase in firmware size and are not directly related to current fmu-v5 and fmuv-6x falling in size.

File: src/drivers/rc_input/ghst_telemetry.cpp (hunk 1 above)
Comment (tonycake) [reply src/drivers/rc_input/ghst_telemetry.cpp:132]: I found this structure, but kind of expected to find some core services to return distance from home, and heading to home, without needing to reproduce the code to compute them.

**Comment by tonycake** [comment]
Ok, formatting changes made. Ready to merge?

**Comment by Igor-Misic** [comment]
> Ok, formatting changes made. Ready to merge?

https://github.com/PX4/PX4-Autopilot/runs/7628282938?check_suite_focus=true

:)

**Review by Igor-Misic — APPROVED** [review]

**Comment by tonycake** [comment]
Many thanks for your help Igor! Much appreciated.

**Pull request merged**